### PR TITLE
Refactors OpenID Connect logic

### PR DESCRIPTION
**Pull request by iredelmeier**
This makes it easier to mock auth flow in the Genie tests.

### CODE REVIEW

File: flux-sdk-common/spec/unit/openid-connect/exchange-credentials-spec.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+          flux_token: 'FLUX_TOKEN',
+        }, {
+          clientSecret: 'CLIENT_SECRET',
+        });
+      });
+
+      it('should request the credentials from the server', function(done) {
+        this.result
+          .then(() => {
+            expect(base64.base64Encode).toHaveBeenCalledWith('CLIENT_ID:CLIENT_SECRET');
```
Comment (daishiflux) [inline flux-sdk-common/spec/unit/openid-connect/exchange-credentials-spec.js]: I still think that this is an implementation detail that makes this test brittle.

File: flux-sdk-common/src/open-id-connect/exchange-credentials.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    headers: {
+      Authorization: `Basic ${encodedAuth}`,
+    },
+  }, { fluxUrl: options.fluxUrl || '' });
+}
+
+/* eslint-disable no-unused-vars */
+function retrieveImplicitCredentials({ code, state, flux_token, ...others }) {
+  return Promise.resolve(others);
```
Comment (daishiflux) [inline flux-sdk-common/src/open-id-connect/exchange-credentials.js]: Instead of using this pattern to remove the fields that you don't want, I think it would be better to confirm that the fields that you expect actually exist.

File: flux-sdk-common/src/open-id-connect/get-authorize-url.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    client_id: clientId,
+    redirect_uri: redirectUri,
+    response_type: responseType,
+    scope: SCOPE,
+  });
+
+  return joinUrl(fluxUrl || FLUX_URL, AUTHORIZE_PATH, search);
+}
+
+function getAuthorizeUrl(state, nonce, clientId, redirectUri, options = {}) {
```
Comment (daishiflux) [inline flux-sdk-common/src/open-id-connect/get-authorize-url.js]: nit: Why do the clientId, redirectUri argument order differ between here and createAuthorizeUrl above?

File: flux-sdk-common/src/open-id-connect/handle-credentials.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  const credentials = {
+    clientId,
+    fluxToken,
+    idToken,
+    scope,
+    accessToken: access_token,
+    refreshToken: refresh_token,
+    tokenExpiry: idToken.payload.exp,
+    tokenType: token_type,
+    clientInfo: {
```
Comment (daishiflux) [inline flux-sdk-common/src/open-id-connect/handle-credentials.js]: Why is this information here? ClientInfo isn't related to OIDC? (There are other fields here that also seem unrelated?)

File: flux-sdk-common/spec/unit/openid-connect/handle-credentials-spec.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        tokenType: 'TOKEN_TYPE',
+        clientInfo: {
+          ClientId: 'CLIENT_ID',
+          ClientVersion: '',
+          AdditionalClientData: {
+            HostProgramVersion: 'unknown',
+            HostProgramMainFile: 'web',
+          },
+          SDKName: 'Flux Javascript SDK',
+          SDKVersion: '0.3.0',
```
Comment (daishiflux) [inline flux-sdk-common/spec/unit/openid-connect/handle-credentials-spec.js]: I would suggest breaking out the SDK version into a constant somewhere to facilitate incrementing versions.

File: flux-sdk-common/spec/unit/openid-connect/exchange-credentials-spec.js (hunk 1 above)
Comment (iredelmeier) [inline flux-sdk-common/spec/unit/openid-connect/exchange-credentials-spec.js]: flux-sdk-common is environment agnostic and therefore doesn't have access to its own base64 encoder. Also, using encoded strings in tests would be much less readable.

File: flux-sdk-common/src/open-id-connect/handle-credentials.js (hunk 4 above)
Comment (iredelmeier) [inline flux-sdk-common/src/open-id-connect/handle-credentials.js]: ClientInfo isn't related to OIDC, but OIDC is related to ClientInfo - currently with respect to the client ID, but possibly with respect to more fields in the future depending on how we handle app names, etc. in the future.

Beyond that, this PR is purely about refactoring existing behaviour. Changing how we handle ClientInfo would be well beyond that scope.

File: flux-sdk-common/spec/unit/openid-connect/handle-credentials-spec.js (hunk 5 above)
Comment (iredelmeier) [inline flux-sdk-common/spec/unit/openid-connect/handle-credentials-spec.js]: That's done in another PR.

File: flux-sdk-common/spec/unit/openid-connect/exchange-credentials-spec.js (hunk 1 above)
Comment (daishiflux) [inline flux-sdk-common/spec/unit/openid-connect/exchange-credentials-spec.js]: We can address this issue separately later.
